index_test.go

```diff
@@ -0,0 +1,263 @@
+      "mappings":{
+        "properties":{
+          "user":{
+            "type":"keyword"
+          },
+          "message":{
+            "type":"text",
+            "store": true,
+            "fielddata": true
+          },
+          "image":{
+            "type":"keyword"
+          },
+          "created":{
+            "type":"date"
+          },
+          "tags":{
+            "type":"keyword"
+          },
+          "location":{
+            "type":"geo_point"
+          }
+        }
+      }
+    }
+  }`
+	exists, err := client.IndexExists(index).Do(context.TODO())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if !exists {
+		if _, err := client.CreateIndexWithMapping(context.TODO(), index, mapping); err != nil {
+			t.Fatal(err)
+		}
+		exists, err = client.IndexExists(index).Do(context.TODO())
+		if err != nil {
+			t.Fatal(err)
+		}
+		if !exists {
+			t.Errorf("expected index created, but not created")
+		}
+	}
+
+	client.deleteIndex(index)
+}
+
+func TestCreateTemplate(t *testing.T) {
+	client, err := New(elastic.SetURL("http://es01:9200"))
+	if err != nil {
+		t.Fatal(err)
+	}
+	tempName := "tweet-template"
+	defer client.deleteTemplate(tempName)
+
+	const template = `
+  {
+    "index_patterns": ["tweet*"],
+    "settings": {
+      "number_of_shards": 1,
+      "number_of_replicas": 0
+    },
+    "mappings": {
+      "properties": {
+        "user":{
+          "type":"keyword"
+       },
+       "message":{
+         "type":"text",
+         "store": true,
+         "fielddata": true
+        },
+        "image":{
+          "type":"keyword"
+        },
+        "created":{
+          "type":"date"
+        },
+        "tags":{
+          "type":"keyword"
+        },
+        "location":{
+          "type":"geo_point"
+        }
+      }
+    }
+  }
+  `
+
+	if _, err := client.CreateTemplate(context.TODO(), tempName, template); err != nil {
+		t.Fatal(err)
+	}
+}
+
+type tweet struct {
+	User     string    `json:"user"`
+	Message  string    `json:"message"`
+	Retweets int       `json:"retweets"`
+	Image    string    `json:"image,omitempty"`
+	Created  time.Time `json:"created,omitempty"`
+	Tags     []string  `json:"tags,omitempty"`
+	Location string    `json:"location,omitempty"`
+}
+
+func TestBulkInsert(t *testing.T) {
+	client, err := New(elastic.SetURL("http://es01:9200"))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	index := "tweets"
+	defer client.Stop()
+
+	_, err = client.Create(context.TODO(), index)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tweets := []tweet{
+		tweet{User: "user1", Message: "message1", Retweets: 1, Image: "image1", Created: time.Now(), Tags: []string{"tag1", "tag2"}, Location: "Tokyo"},
+		tweet{User: "user2", Message: "message2", Retweets: 2, Image: "image2", Created: time.Now(), Tags: []string{"tag1", "tag2"}, Location: "Tokyo"},
+	}
+
+	docs := make([]interface{}, len(tweets))
+	for i, v := range tweets {
+		docs[i] = v
+	}
+
+	bulkRes, err := client.BulkInsert(context.TODO(), index, docs)
```

Comment:
いえ、tweets をそのまま渡せますよ！

あと、察するにinterface{} の理解が十分でないと思いますので、一度確認をおねがいします。

Reply:
tweetsをそのまま渡してエラーになったので、調査したところ以下のドキュメント部分が出てきたので参考にしていました。
しかしinterface{}はどんな型でも受け入れてくれるはずなので、疑問に思っていました。

> Not directly. It is disallowed by the language specification because the two types do not have the same representation in memory. It is necessary to copy the elements individually to the destination slice

https://golang.org/doc/faq#convert_slice_of_interface

> あと、察するにinterface{} の理解が十分でないと思いますので、一度確認をおねがいします。

私も理解が十分でないと感じていますので、確認させていただいます。
もし良い資料などあれば教えていただけないでしょうか？

Reply:
ああ、なるほど。interface{} のスライスはやっかいなんですよね…笑。
解決策のひとつとしては、container list を使えばよいです。
https://golang.org/pkg/container/list/

あとは、リフレクションで動的に型をチェックする、という方法ですかね。
こういうイメージです。
https://qiita.com/ieee0824/items/ef8a5cb79d237f47db68

個人的には前者を好みますが、できそうな方でやってみてください。🙂

Reply:
interface{} が理解できていない訳ではないのであれば大丈夫です。
しいてあげるならこの程度理解できておれば十分なのと、わからないところがあれば別途教えたください！
https://qiita.com/sh-tatsuno/items/0c32c01eaeaf2d726fdf

Reply:
なるほど、いくつかパターンがあるんですね！
こちら確認します！

> interface{} が理解できていない訳ではないのであれば大丈夫です。
しいてあげるならこの程度理解できておれば十分なのと、わからないところがあれば別途教えたください！
https://qiita.com/sh-tatsuno/items/0c32c01eaeaf2d726fdf

ありがとうございます！こちらも確認させていただきます

Reply:
こちら`container/list`を使用する方法で実装してみました。
実装方法について確認していただきたいです！